### ISSUE TITLE
NVDA sometimes queries widgets which have been disposed

### ISSUE THREAD

**Issue by mwhapples** [nvaccess/nvda#11156]
I am involved in developing BrailleBlaster from APH, see www.brailleblaster.org. We have noticed in some development code that if NVDA is running it can cause an exception in our code. It appears that NVDA is making an accessibility query for the caret position after a StyledText widget has been disposed.

Unfortunately this is a development build which I cannot make public, however if someone from NVAccess were to look at this issue we may be able to make it available privately. I have been unable to create a minimal example piece of code replicating this issue. However in searching the internet I came across this SWT bug report https://bugs.eclipse.org/bugs/show_bug.cgi?id=363789 which seems to be a very similar issue, although they refer to having JAWS running. My attempt to replicate the issue using the snippet in the bug report did not throw any exception.

I have replicated our issue with NVDA and our development build of BrailleBlaster on two computers, someone else at APH has also replicated the issue as well.

**Comment by Adriani90** [comment]
Cc: @michaelDCurran

**Comment by feerrenrut** [comment]
There isn't much to go on here. What GUI framework do you use? Are you trying to implement the accessibility API yourself? I assume this is an IAccessible2 call that fails. As I understand, it is the Accessibility API provider's responsibility to manage the life-cycle of objects (see https://docs.microsoft.com/en-gb/windows/win32/winauto/creating-proxy-objects). This is just guesswork as to what might be causing the problem.

@mwhapples you refer to SWT, the Standard Widget Toolkit for Java from Eclipse. I'm guessing you are using this in your application, if so I encourage you to [create a new bug report with them](https://www.eclipse.org/swt/bugs.php). Please cross reference with this issue.

**Comment by mwhapples** [comment]
To answer your questions specifically, although they were answered in the original report.
GUI toolkit, Eclipse SWT.
Is Accessibility API implementation: Relying on the built in SWT accessibility, which in turn uses native controls and I believe uses IAccessible on Windows.
I referred to an Eclipse bug report which seems to relate to a similar issue. However within those comments it seemed like it was suggested that SWT disposes controls in such a way that accessibility tools should be aware of the control being disposed.
My conclusion from the exception traces and what I have read (including the SWT/Eclipse bug report) the issue is caused by screen readers querying a control after it has been disposed, something they should not do. Is it possible that NVDA may be hanging on to a reference to a control/widget longer than it should?

**Comment by jcsteh** [comment]
> it seemed like it was suggested that SWT disposes controls in such a way that accessibility tools should be aware of the control being disposed.

There is EVENT_OBJECT_DESTROY, but in practice, clients don't use this for various reasons. Aside from anything else, there would be potential race conditions for out-of-process clients.

COM rules dictate that a server must not destroy an object until all references are released. Method calls on the object are of course allowed to fail, but they must not cause a crash. Whether NVDA is holding onto a reference longer than SWT would like is irrelevant; destroying the object before NVDA releases it is a violation of COM rules.

One simple thing SWT could do is call CoDisconnectObject on the COM object when the Accessible is disposed. That would forceably disconnect remote clients, releasing any associated references, and prevent further remote calls from reaching the object.

**Comment by mwhapples** [comment]
Thanks for this information, it gives me something to go on when I discuss it with SWT developers or if I were to look into the SWT source code.

**Comment by Adriani90** [comment]
@mwhapples are you still having this problem in NVDA 2024.3? Or did Jamie's comment already clarify your questions? In this case we should close this issue.

**Comment by Adriani90** [comment]
We didn't get an update from the initial author upon request since more than 4 years. I am closing as abandoned. If you are still having this issue, please comment and we can reopen.
